File: admin/app/admin/controller/FaqController.php
<?php
/*
 * Copyright (c) 2026 erik <erik@erik.xyz> — https://erik.xyz
 */

declare(strict_types=1);

namespace app\admin\controller;

use app\model\Faq;
use support\Request;
use support\Response;

class FaqController extends BaseController
{
    /**
     * FAQ 列表
     */
    public function index(Request $request): Response
    {
        $page    = (int) $request->input('page', 1);
        $limit   = (int) $request->input('limit', 15);
        $keyword = $request->input('keyword', '');
        $status  = $request->input('status');

        $query = Faq::query();
        if ($keyword) {
            $query->where('title', 'like', "%{$keyword}%");
        }
        if ($status !== null && $status !== '') {
            $query->where('status', (int) $status);
        }

        $total = $query->count();
        $list  = $query->offset(($page - 1) * $limit)
                       ->limit($limit)
                       ->orderBy('sort', 'asc')
                       ->orderBy('id', 'desc')
                       ->get()
                       ->map(fn($f) => $this->encodeIds($f->toArray()));

        return $this->success([
            'list'  => $list,
            'total' => $total,
            'page'  => $page,
            'limit' => $limit,
        ]);
    }

    /**
     * 新增 FAQ
     */
    public function store(Request $request): Response
    {
        $validator = validator($request->all(), [
            'title'   => 'required|string|max:200',
            'content' => 'required|string',
        ]);

        if ($validator->fails()) {
            return $this->fail($validator->errors()->first(), 422);
        }

        $faq = new Faq();
        $faq->id      = (string) $this->generateId();
        $faq->title   = $request->input('title');
        $faq->content = $request->input('content');
        $faq->sort    = (int) $request->input('sort', 0);
        $faq->status  = (int) $request->input('status', 1);
        $faq->save();

        return $this->success($this->encodeIds($faq->toArray()), '创建成功');
    }

    /**
     * FAQ 详情
     */
    public function show(Request $request, string $hashid): Response
    {
        $id  = $this->decodeId($hashid);
        $faq = Faq::find($id);
        if (!$faq) {
            return $this->fail('FAQ不存在', 404);
        }

        return $this->success($this->encodeIds($faq->toArray()));
    }

    /**
     * 更新 FAQ
     */
    public function update(Request $request, string $hashid): Response
    {
        $id  = $this->decodeId($hashid);
        $faq = Faq::find($id);
        if (!$faq) {
            return $this->fail('FAQ不存在', 404);
        }

        $fillable = ['title', 'content', 'sort', 'status'];
        foreach ($fillable as $field) {
            if ($request->has($field)) {
                $value = $request->input($field);
                if (in_array($field, ['sort', 'status'])) {
                    $value = (int) $value;
                }
                $faq->{$field} = $value;
            }
        }
        $faq->save();

        return $this->success($this->encodeIds($faq->toArray()), '更新成功');
    }

    /**
     * 删除 FAQ
     */
    public function destroy(Request $request, string $hashid): Response
    {
        $id  = $this->decodeId($hashid);
        $faq = Faq::find($id);
        if (!$faq) {
            return $this->fail('FAQ不存在', 404);
        }

        $adminId = $request->adminId ?? 0;
        $error   = $this->confirmPassword($adminId, $request->input('password', ''), $request);
        if ($error !== null) {
            return $this->fail($error, 422);
        }

        $faq->delete();
        return $this->success([], '删除成功');
    }

    /**
     * 批量排序
     */
    public function sortAll(Request $request): Response
    {
        $sorts = $request->input('sorts', []);
        if (empty($sorts) || !is_array($sorts)) {
            return $this->fail('排序数据不能为空', 422);
        }

        foreach ($sorts as $item) {
            try {
                $id  = $this->decodeId($item['id'] ?? '');
                $faq = Faq::find($id);
                if ($faq) {
                    $faq->sort = (int) ($item['sort'] ?? 0);
                    $faq->save();
                }
            } catch (\InvalidArgumentException $e) {
                continue;
            }
        }

        return $this->success([], '排序更新成功');
    }
}

File: admin/apps/flutter/lib/app/pages/faq/faq_controller.dart
/*
 * Copyright (c) 2026 erik <erik@erik.xyz> — https://erik.xyz
 */

import 'package:get/get.dart';
import '../../services/api_service.dart';

/// FAQ 管理控制器
/// 契约对应 admin/app/admin/controller/FaqController.php：
///   GET    /admin/faqs?page&limit&keyword&status
///   POST   /admin/faqs {title, content, sort, status}
///   PUT    /admin/faqs/{id}（title/content/sort/status）
///   DELETE /admin/faqs/{id} {password}（需管理员密码二次确认）
/// 后端无分类字段（列表仅 keyword/status 筛选）；sortAll 端点未注册路由，前端不调用。
/// 列表项：id(hashid), title, content, sort, status, created_at, updated_at
class FaqController extends GetxController {
  final api = ApiService();

  final faqs = <dynamic>[].obs;
  final isLoading = false.obs;
  final total = 0.obs;
  final page = 1.obs;
  final limit = 15.obs;
  final keyword = ''.obs;
  final statusFilter = Rx<int?>(null);

  @override
  void onInit() {
    super.onInit();
    loadFaqs();
  }

  Future<void> loadFaqs({bool reset = false}) async {
    if (reset) page.value = 1;
    isLoading.value = true;
    try {
      final params = <String, dynamic>{
        'page': page.value,
        'limit': limit.value,
      };
      if (keyword.value.isNotEmpty) params['keyword'] = keyword.value;
      if (statusFilter.value != null) params['status'] = statusFilter.value;

      final resp = await api.get('/admin/faqs', params: params);
      faqs.value = resp['data']['list'] as List<dynamic>;
      total.value = resp['data']['total'] as int;
    } catch (e) {
      Get.snackbar('错误', '加载 FAQ 列表失败: $e');
    } finally {
      isLoading.value = false;
    }
  }

  Future<void> search(String kw) async {
    keyword.value = kw.trim();
    await loadFaqs(reset: true);
  }

  Future<void> filterByStatus(int? status) async {
    statusFilter.value = status;
    await loadFaqs(reset: true);
  }

  Future<void> nextPage() async {
    if (page.value * limit.value < total.value) {
      page.value++;
      await loadFaqs();
    }
  }

  Future<void> prevPage() async {
    if (page.value > 1) {
      page.value--;
      await loadFaqs();
    }
  }

  Future<bool> deleteFaq(String id, String password) async {
    try {
      await api.delete('/admin/faqs/$id', data: {'password': password});
      await loadFaqs();
      Get.snackbar('成功', 'FAQ 删除成功');
      return true;
    } catch (e) {
      Get.snackbar('错误', '删除失败: $e');
      return false;
    }
  }
}
